### ISSUE TITLE
Would it be possible to integrate an attention sink https://arxiv.org/pdf/2309.17453.pdf into RetNet?

### ISSUE THREAD

**Issue by pkpro** [syncdoth/RetNet#27]
Apologies, this is not an issue per se; it's a question regarding the possibility of integrating an attention sink from this paper https://arxiv.org/pdf/2309.17453.pdf into RetNet. I'm unable to determine if this will be feasible, given that RetNet contains both the attention and retention mechanisms and there are several forwarding mechanisms. How would the attention sink impact the memory usage if it is possible to integrate with RetNet?

**Comment by syncdoth** [comment]
It's welcome to use issues for discussion 😃

I skimmed through the paper, and the impression I got is: this is special BOS token (attention sink) appended version of window attention. The motivation is to include initial token in the context, which for some reason is important for attention.

I don't think we need this for retention.

1. We don't do softmax, but use deterministic decay + retention.

2. During inference, we can just use recurrent forward, which is just O(T) (attention sink is O(TL))

3. During training, if we have very long sequence, we could use chunkwise forward. 

Parallel forward - O(T^2)
Chunkwise forward - O(TL)
Attention Sink - O(TL)

So in terms of complexity, chunkwise forward should already solve long sequence inefficiency.

**Comment by pkpro** [comment]
Thank you very much for the answer. As I understood the paper the key observation is that the leading tokens getting unproportionately high attention and it is irrelevant what actually those tokens are, the position of those tokens is what curious. They promise 22.2x speedup and up to 4M tokens with this sink attention (of course at expense of the VRAM usage). I'm just curious if this high attention to leading tokens may be the case in the RetNet. I actually do not understand why this high attention to leading tokens is happening. Describing it as a trivial outcome of attention dependency on the same and previous token does NOT exactly mean that leading tokens MUST get accumulated attention... Sorry. I can't put it any better way. Please feel free to close this issue.

**Comment by jploski** [comment]
This "attention sink" (i.e. keeping the initial prompt while throwing away contents of the KV cache when it becomes full) has been a feature in llama.cpp since March 2023 (parameter n_keep), so I doubt the paper authors have really discovered anything new. And I'm not sure why they are so puzzled about why the models pay extra attention to the initial tokens - as these tokens usually contain the instructions during training of instruct models, so it makes perfect sense that the models would focus their attention there (guided by positional embedding).

When it comes to RetNet, which does not have a periodically recycled KV cache, but rather recurrently recomputes the hidden state with each new token, it is a valid question whether a similar problem would occur and whether it could be solved by keeping the initial states more heavily weighed than the later ones. I think it could technically be achieved by some sort of boosting of the initial states and while averaging them into new ones, so as to preven them from decaying at the same rate as those later states.

**Comment by pkpro** [comment]
>> as these tokens usually contain the instructions during training of instruct models, so it makes perfect sense that the models would focus their attention there (guided by positional embedding).

Oh, this makes sense, this could be the RoPE influence.

>> When it comes to RetNet, which does not have a periodically recycled KV cache, but rather recurrently recomputes the hidden state with each new token, it is a valid question whether a similar problem would occur

This would be indeed interesting to investigate. Thank you @jploski.
